src/app/administrator/admin-book/admin-book.component.ts

```diff
@@ -1,15 +1,94 @@
-import { Component, OnInit } from '@angular/core';
+import { Component, OnInit } from "@angular/core";
+import { OLmsService } from "src/app/common/o/services/oLms.service";
+import { environment } from "src/environments/environment";
 
 @Component({
-  selector: 'app-admin-book',
-  templateUrl: './admin-book.component.html',
-  styleUrls: ['./admin-book.component.css']
+  selector: "app-admin-book",
+  templateUrl: "./admin-book.component.html",
+  styleUrls: ["./admin-book.component.css"],
 })
 export class AdminBookComponent implements OnInit {
+  books: any[];
+  authors: any[];
+  genres: any[];
+  publishers: any[];
 
-  constructor() { }
+  constructor(private lmsService: OLmsService) {}
 
   ngOnInit() {
+    this.readBooks();
+    this.readAuthors();
+    this.readGenres();
+    this.readPublishers();
   }
 
+  readBooks() {
+    this.lmsService
+      .get(environment.adminBackendUrl + environment.readBookUri)
+      .subscribe(
+        (result: object[]) => (this.books = result),
+        (error: any) => {
+          // do something with a logger here
+          this.books = [];
+          alert(error.error);
+        }
+      );
+  }
+
+  readAuthors() {
+    this.lmsService
+      .get(environment.adminBackendUrl + environment.readAuthorUri)
+      .subscribe(
+        (result: object[]) => (this.authors = result),
+        (error: any) => {
+          // do something with a logger here
+          this.authors = [];
+          alert(error.error);
+        }
+      );
+  }
+
+  readGenres() {
+    this.lmsService
+      .get(environment.adminBackendUrl + environment.readGenreUri)
+      .subscribe(
+        (result: object[]) => (this.genres = result),
+        (error: any) => {
+          // do something with a logger here
+          this.genres = [];
+          alert(error.error);
+        }
+      );
+  }
+
+  readPublishers() {
+    this.lmsService
+      .get(environment.adminBackendUrl + environment.readPublisherUri)
+      .subscribe(
+        (result: object[]) => (this.publishers = result),
+        (error: any) => {
+          // do something with a logger here
+          this.publishers = [];
+          alert(error.error);
+        }
+      );
+  }
+
+  deleteBook(bookId: number) {
+    if (!confirm("Delete this book?")) return;
+    this.lmsService
+      .delete(
+        `${environment.adminBackendUrl}${environment.deleteBookUri}/${bookId}`
+      )
+      .subscribe(null, (error: any) => {
+        // do something with a logger here
+        alert(error.error);
+      })
+      .add(() => {
+        this.readBooks();
+        this.readAuthors();
+        this.readGenres();
+        this.readPublishers();
+      });
+  }
```

Comment:
What is the purpose of this piece of code?

Reply:
It makes sure the HTTP call has already done its thing when information is re-read from the database, which fixes a bug I ran into where the view wasn't always updating correctly when changes were made.  I need to add it to the override and publisher components too; I never observed the bug with them, probably because their backend stuff is simpler and faster.

Reply:
Also, using add() makes it so the code gets executed whether there's an error or not.